### PR TITLE
Fixing issue where UIColor with alpha < 1 was being rendered incorrectly

### LINKED ISSUE #125: UIColor with alpha < 1 is rendered incorrectly
Fill color with alpha < 1 is rendered incorrectly when it is evaluated from `NSExpression`
The issue is a regression after fixing mapbox/mapbox-gl-native#14329 in mapbox/mapbox-gl-native#14406

Color components are premultiplied twice in some cases:
The first time in `UIColor+MGLAdditions.mm`
```
- (mbgl::Color)mgl_color
{
    CGFloat r, g, b, a;
    [self getRed:&r green:&g blue:&b alpha:&a];
    // UIColor provides non-premultiplied color components, so we have to premultiply each
    // color component with the alpha value to transform it into a valid
    // mbgl::Color which expects premultiplied color components.
    return { static_cast<float>(r*a), static_cast<float>(g*a), static_cast<float>(b*a), static_cast<float>(a) };
}
```
Then in `util.cpp`
```
Result<Color> rgba(double r, double g, double b, double a) {
    if (
        r < 0 || r > 255 ||
        g < 0 || g > 255 ||
        b < 0 || b > 255
    ) {
        return EvaluationError {
            "Invalid rgba value [" + stringifyColor(r, g, b, a)  +
            "]: 'r', 'g', and 'b' must be between 0 and 255."
        };
    }
    if (a < 0 || a > 1) {
        return EvaluationError {
            "Invalid rgba value [" + stringifyColor(r, g, b, a)  +
            "]: 'a' must be between 0 and 1."
        };
    }
    return Color(r / 255 * a, g / 255 * a, b / 255 * a, a);
}
```
### Steps to reproduce

 1. Apply following patch to MBXViewController.m
```
diff --git a/platform/ios/app/MBXViewController.m b/platform/ios/app/MBXViewController.m
index 82a68e074..d9d7cfcde 100644
--- a/platform/ios/app/MBXViewController.m
+++ b/platform/ios/app/MBXViewController.m
@@ -1084,9 +1084,14 @@ CLLocationCoordinate2D randomWorldCoordinate() {
     [self.mapView.style addSource:source];
 
     MGLFillStyleLayer *fillLayer = [[MGLFillStyleLayer alloc] initWithIdentifier:@"test" source:source];
-    fillLayer.fillColor = [NSExpression expressionForConstantValue:[UIColor purpleColor]];
+    UIColor *color = [[UIColor blueColor] colorWithAlphaComponent:0.5];
+    fillLayer.fillColor = [NSExpression expressionWithFormat:@"MGL_MATCH(MGL_FUNCTION(\'get\', \'type\'),\'park\', %@, %@)", color, color];
+    fillLayer.fillOpacity = [NSExpression expressionForConstantValue:@1.0];
     [self.mapView.style addLayer:fillLayer];
-
+    
+    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 100, 100)];
+    view.backgroundColor = color;
+    [self.mapView addSubview:view];
 }
 
 - (void)styleSymbolLayer
```
 2. Build and run the app
 3. Select `Add Shapes & Apply fill`
 4. Navigate to Amsterdam

### Expected behavior

Native UIView with 0.5 opacity and blue color should match what is rendered by mapbox

### Actual behavior
Color is different in comparison to native UIView

Screen Shot 2020-01-02 at 3 32 30 PM

### Configuration

**Mapbox SDK versions:** 5.4/5.5
**iOS/macOS versions:** iOS 12.2
**Device/simulator models:** iOS Simulator 5s
**Xcode version:** 11.1

* I can't try on the latest release due to compilation error

### PR DESCRIPTION
**Pull request by nishant-karajgikar**
`<changelog> Fixed an issue where properties such as MGLFillStyleLayer.fillColor and MGLLineStyleLayer.lineColor misinterpreted non-opaque UIColors.</changelog>`

Fixes #125 

This issue was being caused due to an incorrect assumption that the R, G, B color components needed to be multiplied with the alpha value. The changes fix this by using the newly exposed `mgl_colorForPremultipliedValue` category method on `UIColor`.

### CODE REVIEW

File: platform/ios/src/UIColor+MGLAdditions.mm [hunk 1]
```diff
@@ -12,6 +12,14 @@ @implementation UIColor (MGLAdditions)
     return { static_cast<float>(r*a), static_cast<float>(g*a), static_cast<float>(b*a), static_cast<float>(a) };
 }
 
+// Intended to be used when providing colors on the basis of an NSExpression
+- (mbgl::Color)mgl_colorForPremultipliedValue
```
Comment (1ec5) [inline platform/ios/src/UIColor+MGLAdditions.mm:16]: The two remaining code paths using `-mgl_color` are for styling annotations:

* `-[MGLMapView fillColorForPolygonAnnotation:]` → `-[MGLPolygon annotationObjectWithDelegate:]` → `FillAnnotationImpl::updateStyle()` → `FillLayer::setFillColor()`
* `-[MGLMapView strokeColorForShapeAnnotation:]` → `-[MGLPolyline annotationObjectWithDelegate:]` → `LineAnnotationImpl::updateStyle()` → `LineLayer::setLineColor()`

Annotations are backed by style layers under the hood, but the colors are expressed as raw `mbgl::PropertyValue<mbgl::Color>`s rather than expressions. Is the expectation of premultiplied values specific to NSExpression or paint property values in general?

In other words, do shape annotations have the correct colors when the application specifies the same UIColors as they would in style layer properties?

/ref https://github.com/mapbox/mapbox-gl-native/pull/16223#discussion_r383694089

Comment (fabian-guerra) [reply platform/ios/src/UIColor+MGLAdditions.mm:16]: > Annotations are backed by style layers under the hood, but the colors are expressed as raw mbgl::PropertyValue<mbgl::Color>s rather than expressions. Is the expectation of premultiplied values specific to NSExpression or paint property values in general?

Expressions expect non-premultiplied, and `mbgl::Color` premultiplied. According to this comment https://github.com/mapbox/mapbox-gl-native/pull/16153#issuecomment-580171975

File: platform/darwin/test/MGLExpressionTests.mm [hunk 2]
```diff
@@ -296,7 +296,7 @@ - (void)testConstantValueExpressionObject {
         float blue = (213.0 * alpha) / 255;
         MGLColor *color = [MGLColor mgl_colorWithColor:{ red, green, blue, alpha }]; // papayawhip
         NSExpression *expression = [NSExpression expressionForConstantValue:color];
-        NSArray *jsonExpression = @[@"rgba", @127.5, @119.5, @106.5, @0.5];
+        NSArray *jsonExpression = @[@"rgba", @255.0, @239.0, @213.0, @0.5];
         XCTAssertEqualObjects(expression.mgl_jsonExpressionObject, jsonExpression);
         XCTAssertEqualObjects([expression expressionValueWithObject:nil context:nil], color);
```
Comment (1ec5) [inline platform/darwin/test/MGLExpressionTests.mm:301]: This tests that Apple’s NSExpression code can get a constant value out of a constant value expression, but it doesn’t test that a JSON `rgba` expression gets converted to the right NSColor or UIColor. Can we add a test to that effect? I’m concerned that we’re only fixing the bug in one direction but not the other.

**Comment by nishant-karajgikar** [comment]
> please rebase onto master and remove the mbgl submodule bump

Will do, it's weird that there's a change in the `mbgl` submodule. I didn't make any changes to that.

File: platform/darwin/test/MGLExpressionTests.mm (hunk 2 above)
Comment (nishant-karajgikar) [reply platform/darwin/test/MGLExpressionTests.mm:301]: Just added a test that hopefully covers your concern!

File: platform/darwin/test/MGLExpressionTests.mm [hunk 3]
```diff
@@ -300,6 +300,28 @@ - (void)testConstantValueExpressionObject {
...
+        float green = (239.0 * alpha) / 255;
+        float blue = (213.0 * alpha) / 255;
+
+        MGLColor *color = [MGLColor mgl_colorWithColor:{ red, green, blue, alpha }]; // papayawhip
+        CGFloat originalRed = 0.0, originalGreen = 0.0, originalBlue = 0.0, originalAlpha = 0.0;
+        [color getRed:&originalRed green:&originalGreen blue:&originalBlue alpha:&originalAlpha];
+
+        NSArray *jsonExpression = @[@"rgba", @255.0, @239.0, @213.0, @0.5];
+        NSExpression *papayawhipExpression = [NSExpression expressionWithMGLJSONObject:jsonExpression];
+        MGLColor *papayaWhipColor = [papayawhipExpression expressionValueWithObject:nil context:nil];
```
Comment (1ec5) [inline platform/darwin/test/MGLExpressionTests.mm:316]: By the way, `papayawhipExpression` is a constant value expression, so this is equivalent to getting its `constantValue`.

File: platform/darwin/test/MGLExpressionTests.mm [hunk 4]
```diff
@@ -300,6 +300,28 @@ - (void)testConstantValueExpressionObject {
...
+    {
+        // Transform color components to non-premultiplied values
+        float alpha = 0.5;
+        float red = (255.0 * alpha) / 255;
+        float green = (239.0 * alpha) / 255;
+        float blue = (213.0 * alpha) / 255;
+
+        MGLColor *color = [MGLColor mgl_colorWithColor:{ red, green, blue, alpha }]; // papayawhip
+        CGFloat originalRed = 0.0, originalGreen = 0.0, originalBlue = 0.0, originalAlpha = 0.0;
+        [color getRed:&originalRed green:&originalGreen blue:&originalBlue alpha:&originalAlpha];
```
Comment (1ec5) [inline platform/darwin/test/MGLExpressionTests.mm:312]: This code takes three components, creates an `mbgl::Color`, converts it to an NSColor/UIColor, then gets the components back out, calling them “originals” but doesn’t check that they match the original `float`s defined above.

File: platform/darwin/test/MGLExpressionTests.mm [hunk 5]
```diff
@@ -300,6 +300,28 @@ - (void)testConstantValueExpressionObject {
...
+
+        NSArray *jsonExpression = @[@"rgba", @255.0, @239.0, @213.0, @0.5];
+        NSExpression *papayawhipExpression = [NSExpression expressionWithMGLJSONObject:jsonExpression];
+        MGLColor *papayaWhipColor = [papayawhipExpression expressionValueWithObject:nil context:nil];
+        CGFloat convertedRed = 0.0, convertedGreen = 0.0, convertedBlue = 0.0, convertedAlpha = 0.0;
+        [papayaWhipColor getRed:&convertedRed green:&convertedGreen blue:&convertedBlue alpha:&convertedAlpha];
+
+        XCTAssertEqualWithAccuracy(convertedRed, originalRed, 0.0000001);
+        XCTAssertEqualWithAccuracy(convertedGreen, originalGreen, 0.0000001);
+        XCTAssertEqualWithAccuracy(convertedBlue, originalBlue, 0.0000001);
```
Comment (1ec5) [inline platform/darwin/test/MGLExpressionTests.mm:322]: I think it would be easier to sight-verify that this test is correct if the assertions tested equality to the elements of `jsonExpression` instead of these “original” variables that have already been round-tripped:

```objc
NSArray *jsonExpression = @[@"rgba", @255.0, @239.0, @213.0, @0.5];
MGLColor *papayaWhipColor = [NSExpression expressionWithMGLJSONObject:jsonExpression].constantValue;
CGFloat convertedRed = 0.0, convertedGreen = 0.0, convertedBlue = 0.0, convertedAlpha = 0.0;
[papayaWhipColor getRed:&convertedRed green:&convertedGreen blue:&convertedBlue alpha:&convertedAlpha];
XCTAssertEqualWithAccuracy(convertedRed, 255, 0.0000001);
XCTAssertEqualWithAccuracy(convertedGreen, 239, 0.0000001);
XCTAssertEqualWithAccuracy(convertedBlue, 213, 0.0000001);
XCTAssertEqual(convertedAlpha, 0.5);
```

This shorter version still provides coverage of `+[MGLColor mgl_colorWithColor:]`, just not directly.

Comment (nishant-karajgikar) [reply platform/darwin/test/MGLExpressionTests.mm]: Got it, thanks for the input! I've made the changes to the test!

File: platform/ios/CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,12 @@
 
 Mapbox welcomes participation and contributions from everyone. Please read [CONTRIBUTING.md](../../CONTRIBUTING.md) to get started.
 
+## 5.9.0
+
+### Styles and rendering
+
+* Fixed an issue where properties such as `MGLFillStyleLayer.fillColor` and `MGLLineStyleLayer.lineColor` misinterpreted non-opaque `UIColor`s.
```
Comment (1ec5) [inline platform/ios/CHANGELOG.md:9]:
```suggestion
* Fixed an issue where properties such as `MGLFillStyleLayer.fillColor` and `MGLLineStyleLayer.lineColor` misinterpreted non-opaque `UIColor`s. ([#266](https://github.com/mapbox/mapbox-gl-native-ios/pull/266))
```

File: platform/macos/CHANGELOG.md [hunk 7]
```diff
@@ -23,6 +23,7 @@
 * Fixed an issue where style layers backed by a shape source could flicker when transitioning between styles. ([#15907](https://github.com/mapbox/mapbox-gl-native/pull/15907), [#15941](https://github.com/mapbox/mapbox-gl-native/pull/15941))
 * Improved the performance of loading a style that has many style images. ([#16187](https://github.com/mapbox/mapbox-gl-native/pull/16187))
 * Updated “map ID” to the more accurate term “tileset ID” in documentation; updated “style's Map ID” to the more accurate term “style URL”. ([#15116](https://github.com/mapbox/mapbox-gl-native/pull/15116))
+* Fixed an issue where properties such as `MGLFillStyleLayer.fillColor` and `MGLLineStyleLayer.lineColor` misinterpreted non-opaque `NSColor`s.
```
Comment (1ec5) [inline platform/macos/CHANGELOG.md:26]:
```suggestion
* Fixed an issue where properties such as `MGLFillStyleLayer.fillColor` and `MGLLineStyleLayer.lineColor` misinterpreted non-opaque `NSColor`s. ([#266](https://github.com/mapbox/mapbox-gl-native-ios/pull/266))
```

File: platform/ios/src/UIColor+MGLAdditions.mm (hunk 1 above)
Comment (1ec5) [reply platform/ios/src/UIColor+MGLAdditions.mm:16]: mapbox/mapbox-gl-native-ios#454
